File: leetcode/k_closest_points_to_origin/k_closest_points_to_origin.java
package k_closest_points_to_origin;

import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

class Solution {
    class Point {
        private int x;
        private int y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        int distanceSquare() {
            return x*x + y*y;
        }
    }

    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<Point> pq = new PriorityQueue<>(
            (a,b) -> Integer.compare(b.distanceSquare(), a.distanceSquare())
        );

        for(int[] p : points) {
            pq.add(new Point(p[0], p[1]));
            if(pq.size() > k) {
                pq.poll();
            }
        }

        List<int[]> ans = new ArrayList<>();
        while(!pq.isEmpty()) {
            Point p = pq.poll();
            ans.add(new int[]{p.x, p.y});
        }

        return ans.toArray(new int[0][]);
    }
}

File: leetcode/k_closest_points_to_origin/k_closest_points_to_origin.cpp
/*
 * Time Complexity: O(N log K)
 *   - N is the number of points, K is the number to return.
 *   - Each insertion and removal from the heap is O(log K).
 *
 * Space Complexity: O(K)
 *   - For the heap storing K points.
 */

#include<vector>
#include<queue>

using std::vector;
using std::priority_queue;

struct Point {
    int x;
    int y;

    Point(int _x, int _y) : x(_x), y(_y) {}
};

struct CompareByDist {
    bool operator()(const Point &a, const Point &b) const {
        int distA = a.x * a.x + a.y * a.y;
        int distB = b.x * b.x + b.y * b.y;
        return distA < distB; 
    }
};

class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        priority_queue<Point, vector<Point>, CompareByDist> pq;

        for(vector<int> p : points) {
            pq.emplace(p[0], p[1]);

            if(pq.size() > k) {
                pq.pop();
            }
        }
        
        vector<vector<int>> ans;
        while(!pq.empty()) {
            Point p = pq.top();
            ans.push_back({p.x,p.y});
            pq.pop();
        }

        return ans;
    }
};